### ISSUE TITLE
Evaluation code

### ISSUE THREAD

**Issue by zoezhou1999** [xg-chu/GAGAvatar#43]
Hi @xg-chu 
Thank you for sharing your amazing work!
Is it possible for you to share evaluation code for fair comparison?

Thank you!

Best
Yuhongze

**Comment by fusheng-ji** [comment]
Hi Xuangeng, thank you for your awesome work!

I will be appreciate if you could share your evaluation script.

**Comment by xg-chu** [comment]
Thank you for your interest to our work! ^_^
Here are some codes for metrics.
Deep3DEngine is a wrapper for Deep3DFaceRecon, landmarks are the 68 landmarks from insightface, and the insight face model is also used to preprocess the input of deep3dfacerecon (for the face_align.norm_crop).

```
class ExpressionEngine():
    def __init__(self, device='cuda'):
        self.deep3d_engine = Deep3DEngine(with_emoca=False, device=device, lazy_init=False)
        self.arcface = self.deep3d_engine.mica_model.arcface

    @torch.no_grad()
    def forward(self, inp, preds, target, data_range=(0.0, 1.0), cross_id=False):
        p_mframe, p_bbox, p_kps, _ = self.deep3d_engine._crop_frame(preds*255.0)
        t_mframe, t_bbox, t_kps, _ = self.deep3d_engine._crop_frame(target*255.0)
        if p_kps is None or t_kps is None:
            return 0, 0, 0, 0, 0
        pred_eres = self.deep3d_engine.process_frame_with_kps(preds*255.0, p_kps)
        target_eres = self.deep3d_engine.process_frame_with_kps(target*255.0, t_kps)
        
        pred_idcode = torch.nn.functional.normalize(self.arcface(p_mframe.cuda()))
        if cross_id:
            i_mframe, _, _, _ = self.deep3d_engine._crop_frame(inp*255.0)
            inp_idcode = torch.nn.functional.normalize(self.arcface(i_mframe.cuda()))
            idsim = torch.nn.functional.cosine_similarity(pred_idcode, inp_idcode)
        else:
            target_idcode = torch.nn.functional.normalize(self.arcface(t_mframe.cuda()))
            idsim = torch.nn.functional.cosine_similarity(pred_idcode, target_idcode)

        aed = torch.nn.functional.l1_loss(pred_eres['exp'], target_eres['exp'])
        apd = torch.nn.functional.l1_loss(pred_eres['pose'][:3], target_eres['pose'][:3])
        return idsim, aed, apd, p_kps, t_kps
```

**Comment by fusheng-ji** [comment]
Hello @xg-chu,

Thank you for sharing the evaluation code clip. 

I have a question about the calculation way of apd loss, If I don't misunderstand, the pred_eres['pose'] and target_eres['pose'] are the concatenation of rotation angle and translation, right? 

So my question is Why do you only take the last 3 dims, which represent translation parameters for loss computation?

Because I have read the [OTAvatar's evaluation method](https://github.com/theEricMa/OTAvatar/issues/27#issuecomment-1801217498). And they calculated the L1 loss on both rotation and translation. 

Can you share with us more details about this part?

Many thanks!

**Comment by xg-chu** [comment]
I use the rotation for pose evaluation, not translation. I think it would be OK to use both for APD as long as all baselines are processed in the same way.

**Comment by aminfadaei116** [comment]
> Thank you for your interest to our work! ^_^ Here are some codes for metrics. Deep3DEngine is a wrapper for Deep3DFaceRecon, landmarks are the 68 landmarks from insightface, and the insight face model is also used to preprocess the input of deep3dfacerecon (for the face_align.norm_crop).
> 
> ```
> class ExpressionEngine():
>     def __init__(self, device='cuda'):
>         self.deep3d_engine = Deep3DEngine(with_emoca=False, device=device, lazy_init=False)
>         self.arcface = self.deep3d_engine.mica_model.arcface
> 
>     @torch.no_grad()
>     def forward(self, inp, preds, target, data_range=(0.0, 1.0), cross_id=False):
>         p_mframe, p_bbox, p_kps, _ = self.deep3d_engine._crop_frame(preds*255.0)
>         t_mframe, t_bbox, t_kps, _ = self.deep3d_engine._crop_frame(target*255.0)
>         if p_kps is None or t_kps is None:
>             return 0, 0, 0, 0, 0
>         pred_eres = self.deep3d_engine.process_frame_with_kps(preds*255.0, p_kps)
>         target_eres = self.deep3d_engine.process_frame_with_kps(target*255.0, t_kps)
>         
>         pred_idcode = torch.nn.functional.normalize(self.arcface(p_mframe.cuda()))
>         if cross_id:
>             i_mframe, _, _, _ = self.deep3d_engine._crop_frame(inp*255.0)
>             inp_idcode = torch.nn.functional.normalize(self.arcface(i_mframe.cuda()))
>             idsim = torch.nn.functional.cosine_similarity(pred_idcode, inp_idcode)
>         else:
>             target_idcode = torch.nn.functional.normalize(self.arcface(t_mframe.cuda()))
>             idsim = torch.nn.functional.cosine_similarity(pred_idcode, target_idcode)
> 
>         aed = torch.nn.functional.l1_loss(pred_eres['exp'], target_eres['exp'])
>         apd = torch.nn.functional.l1_loss(pred_eres['pose'][:3], target_eres['pose'][:3])
>         return idsim, aed, apd, p_kps, t_kps
> ```

Thanks for sharing this. Is it possible for you to also share the repo where the Deep3DEngine/Deep3DFaceRecon class is?

**Comment by JianqiangRen** [comment]
> Hi Xuangeng, thank you for your awesome work!
> 
> I will be appreciate if you could share your evaluation script.

Can you reproduce the lpips result (0.122 on VFHQ dataset) reported in paper?  I cannot reproduce it...
